### ISSUE TITLE
Inline workers does not work in Edge Legacy anymore

### ISSUE THREAD

**Issue by mpk** [webpack-contrib/worker-loader#292]
- Operating System: Windows 10
- Node Version: 12.14.0
- NPM Version: 6.13.4
- webpack Version: 4.44.2
- worker-loader Version: 3.0.3

### Expected Behavior

Inline worker should execute in Edge Legacy

### Actual Behavior

Inline worker does not execute in Edge Legacy

### Code

```js
// webpack.config.js
let HtmlWebpackPlugin = require('html-webpack-plugin');

module.exports = {
	entry: './index.js',
	plugins: [new HtmlWebpackPlugin()]
};
```

```js
// index.js
import TestWorker from 'worker-loader?inline=no-fallback!./worker';

let worker = new TestWorker();
console.log(worker);
```

```js
// worker.js
console.log('Log from worker');
```

### How Do We Reproduce?
The code should write "Log from worker" into the console. However, it does not work in Edge Legacy browser (tested on version 18.18363). The worker is created successfully, but its contents are not executed, so nothing gets logged into the console.

I have looked at the worker-loader code and found out that the problem is in `URL.revokeObjectURL(objectURL);` line in `inline.js`. When I commented out that line, Edge executed the worker contents. Wrapping the call in setTimeout with 1000ms delay worked as well, but that solution seems quite hacky.

This issue is not present in worker-loader 2.0.0

**Comment by evilebottnawi** [comment]
Feel free to send a fix:
```js
if (URL.revokeObjectURL) {
  URL.revokeObjectURL(objectURL);
}
```

Should be easy

**Comment by mpk** [comment]
No, Edge Legacy actually supports URL.revokeObjectURL, but it does not like when the Blob URL is revoked "too quickly".

**Comment by Rush** [comment]
inline workers stopped working for me after upgrading to 3.x and changing `inline: true` to `inline: 'no-fallback'`

![image](https://user-images.githubusercontent.com/115926/94878882-51a47c00-0424-11eb-9ddb-d883b6e49e3c.png)

Investigating network shows a script with contents:
```
importScripts('http://localhost:4000/[object%20Module]');
```

**Comment by evilebottnawi** [comment]
@mpk Can you create reproducible test repo?

**Comment by evilebottnawi** [comment]
@Rush It is not you problem, please provide reproducible test repo, I will help

**Comment by mpk** [comment]
@evilebottnawi Sure, here you go: https://github.com/mpk/worker-loader-edge

**Comment by alexander-akait** [comment]
@mpk to be honestly I can't reproduce :confused:

**Comment by KallynGowdy** [comment]
I'm having a similar issue with Chrome (`87.0.4280.88`) and Edgium (`87.0.664.55`) on worker-loader `v3.0.6` with webpack `v5.10.0`.
It is worth noting that the worker loads fine on Firefox.

Definitely seems like a race condition because if I put a breakpoint at [`inline.js line 31`](https://github.com/webpack-contrib/worker-loader/blob/master/src/runtime/inline.js#L31) and continue after half a second then the worker loads correctly but if I disable the breakpoint then it does not load.

In light of this, the root cause appears to actually be a Chrome bug however I haven't found anything in their bug tracker.

**Comment by Shravan-Joopally** [comment]
I could reproduce this issue consistently on IE 11. Definitely revoking the object url immediately is issue. If I tried below code its working fine.

```
import TestWorker from 'worker-loader?inline=no-fallback!./worker';
let tmp = URL.revokeObjectURL;
URL.revokeObjectURL = function() {};
let worker = new TestWorker();
URL.revokeObjectURL = tmp;
console.log(worker);
```

**Comment by alexander-akait** [comment]
Can you check `revokeObjectURL` is exist?

**Comment by Shravan-Joopally** [comment]
yes its there

**Comment by alexander-akait** [comment]
Weird, it should work, no errors?

**Comment by Shravan-Joopally** [comment]
yeah, no errors

**Comment by alexander-akait** [comment]
Looks bug in browser...

**Comment by lawrence-witt** [comment]
I'm also running into this issue with inline workers in IE11. Changing the URL revocation in `inline.js` to the following fixes it in my use case:

`setTimeout(function() { URL.revokeObjectURL(objectURL) });`

**Comment by alexander-akait** [comment]
Very very very weird, somebody can provide versions and steps to reproduce, I am not against to fix it using `setTimeout`, but I want to investigate it

**Comment by elasim** [comment]
Is there any chance to merge the patch about setTimeout on above or other plan for this issue?

**Comment by alexander-akait** [comment]
@elasim Can you provide screenshot/etc of the problem?

**Comment by elasim** [comment]
seems like worker is forked with empty script.

![image](https://user-images.githubusercontent.com/1837682/117677496-79916480-b1e9-11eb-8775-871690b46b0a.png)
![image](https://user-images.githubusercontent.com/1837682/117677528-80b87280-b1e9-11eb-82f6-d5c8d888c882.png)
![image](https://user-images.githubusercontent.com/1837682/117677844-c8d79500-b1e9-11eb-81dd-0eee7d984a9a.png)

**Comment by elasim** [comment]
> Can you console log https://github.com/webpack-contrib/worker-loader/blob/master/src/runtime/inline.js#L23 `content` and `blob` and show me it here?

Are you sure it is L23 what you want to put it log?
btw, it is  the log of content and blob

![image](https://user-images.githubusercontent.com/1837682/117680317-02a99b00-b1ec-11eb-984e-369627004371.png)


```
/******/ (function() { // webpackBootstrap
/******/ 	"use strict";
/******/ 	// The require scope
/******/ 	var __webpack_require__ = {};
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/define property getters */
/******/ 	!function() {
/******/ 		// define getter functions for harmony exports
/******/ 		__webpack_require__.d = function(exports, definition) {
/******/ 			for(var key in definition) {
/******/ 				if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
/******/ 					Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
/******/ 				}
/******/ 			}
/******/ 		};
/******/ 	}();
/******/ 	
/******/ 	/* webpack/runtime/hasOwnProperty shorthand */
/******/ 	!function() {
/******/ 		__webpack_require__.o = function(obj, prop) { return Object.prototype.hasOwnProperty.call(obj, prop); }
/******/ 	}();
/******/ 	
/******/ 	/* webpack/runtime/make namespace object */
/******/ 	!function() {
/******/ 		// define __esModule on exports
/******/ 		__webpack_require__.r = function(exports) {
/******/ 			if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 				Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 			}
/******/ 			Object.defineProperty(exports, '__esModule', { value: true });
/******/ 		};
/******/ 	}();
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
/*!*************************************************************************************************************************************************************************************************************************************************************!*\
  !*** ../../common/temp/node_modules/.pnpm/babel-loader@8.2.2/node_modules/babel-loader/lib/index.js??ruleSet[1].rules[1].use[0]!../../common/temp/node_modules/.pnpm/source-map-loader@2.0.1/node_modules/source-map-loader/dist/cjs.js!./lib/ie-worker.js ***!
  \*************************************************************************************************************************************************************************************************************************************************************/
__webpack_require__.r(__webpack_exports__);
/* harmony export */ __webpack_require__.d(__webpack_exports__, {
/* harmony export */   "default": function() { return __WEBPACK_DEFAULT_EXPORT__; }
/* harmony export */ });
function handleMessage(event) {
  self.postMessage(event.data);
}

self.addEventListener("message", handleMessage);

var default_1 =
/** @class */
function () {
  function default_1() {}

  return default_1;
}();

/* harmony default export */ const __WEBPACK_DEFAULT_EXPORT__ = (default_1);
/******/ })()
;
```

**Comment by alexander-akait** [comment]
I want to look at blob content here https://github.com/webpack-contrib/worker-loader/blob/master/src/runtime/inline.js#L29, maybey `BlobBuilder` or `getBlob` is broken

**Comment by elasim** [comment]
It's not. BlobBuilder won't used. https://github.com/webpack-contrib/worker-loader/blob/master/src/runtime/inline.js#L14-L25 is never reached because globalScope.Blob is available.

**Comment by alexander-akait** [comment]
Can you show `blob` value here https://github.com/webpack-contrib/worker-loader/blob/master/src/runtime/inline.js#L29, here two places for bug
- blob generation is broken
- `URL.createObjectURL` is broken with blob
- `URL.revokeObjectURL` is broken without any reports

Also do you have CORS or other security headers?

**Comment by elasim** [comment]
I've already attached. see the last line of logs.

![](https://user-images.githubusercontent.com/1837682/117680317-02a99b00-b1ec-11eb-984e-369627004371.png)

- blob generation is broken: No. It's working If I commented out revokeObjectURL or give some delay using setTimeout.
- CORS or Security Header: No. this script served via webpack-dev-server. there are no such a security header.
- URL.createObjectURL is broken with blob: No. As you can see the last image, We can read blob data via xhr and there are no error.
- URL.revokeObjectURL is broken without any reports: Maybe. but no way to figure out.

**Comment by alexander-akait** [comment]
What about `setImmediate`? Can you try? Maybe we can check `setImmediate` is defined and use logic, so for good browser we will keep original logic

**Comment by elasim** [comment]
https://github.com/webpack-contrib/worker-loader/blob/master/src/runtime/inline.js#L35

Replacing that line with `globalScope.setImmediate(URL.revokeObjectURL, objectURL);` works.

**Comment by elasim** [comment]
> Ideally we need `if (globalScope.setImmediate) { globalScope.setImmediate(() => { URL.revokeObjectURL(objectURL) }); } else { URL.revokeObjectURL(objectURL); }`

This one also works.

**Comment by alexander-akait** [comment]
Feel free to send a fix to our runtime if it is work https://github.com/webpack-contrib/worker-loader/blob/master/src/runtime/inline.js#L35, no need tests, just add comment about version and problem

**Comment by alexander-akait** [comment]
Can you clarify?

**Comment by elasim** [comment]
Here is pseudo code to reproduce.

```ts
// inline-worker.js
self.addEventListener('message', ({ data }) => {
  // (IE bug) this line is not work when inlineWorkerUrl was revoked.
  const url = Object.createObjectURL(new Blob(['something']));

  self.postMessage({ url });
});
```

```ts
// main.js
const objectUrl = URL.createObjectUrl(inlineWorkerBlob);
const worker = new Worker(objectUrl);

worker.addEventListener('message', () => {
  URL.revokeObjectUrl(objectUrl);
});

setTimeout(() => {
  // // expect to success, actual also succeed.
  worker.postMessage(1);
}, 1000);
setTimeout(() => {
  // expect to success, actual will fail because of bug
  // objectUrl was revoked at first callback event
  worker.postMessage(2);
}, 2000);
```

**Comment by alexander-akait** [comment]
I don't understand...

**Comment by elasim** [comment]
Okay, maybe I'll capture some video stub with example repo for your understand.

**Comment by alexander-akait** [comment]
What are you trying to achieve? I don't understand at all, here is a browser problem and that's it. We need search approach to fix it.

**Comment by elasim** [comment]
Oh, I was lost completely.
I was telling you what's the issue is.

So, back to how to fix it.
Maybe we got 2 options.

1. Give up self-closing and delegate revoking to the author who people using it.
2. Putting some codes to revoke when it closed. perhaps we need some special message for this and it will need to change original behavior. 

```ts
// Inject stub if worker launched inline worker. (but we can't determine that at compile-time.)
// maybe here? https://github.com/webpack-contrib/worker-loader/blob/22275e9cb0b67bc008ea2b008542303493eede18/src/utils.js#L84
var origTerminate = worker.terminate;
worker.terminate = () => {
  URL.revokeObjectURL(objectUrl);
  return origTerminate.call(worker);
};
worker.addEventListener('message', ({ data }) => {
  if (data === ''$specialMessageForClosing'') {
  }
});
```

```ts
var origClose = self.close;
self.close = () => {
  self.postMessage('$specialMessageForClosing');
  origClose.call(self);
}
```

I think option-1 is reasonable.

**Comment by alexander-akait** [comment]
If you forget to run `URL.revokeObjectURL`, you will have memory leak, If this problem is so complex I would rather say that we do not support workers in Edge Legacy browsers due bug in browsers and you can write workaround on application side.

If you need deeply help, please create reproducible repo, provide version of browser(s) and os information. I will write to Microsoft and ask them how we can fix it.

**Comment by elasim** [comment]
What about adding new option for manual revoke? Is this too complicated?
I'll post reproducible repo anyway.

**Comment by alexander-akait** [comment]
I want to avoid new options if we can solve it on code level

**Comment by elasim** [comment]
I attached reproducible link and test results.
Also, I'm not used webpack or worker-loader on this testbed intentionally for simplifying the issue.

https://elasim.github.io/test/ie-inline-worker-bug

![image](https://user-images.githubusercontent.com/1837682/118139662-8c52a600-b442-11eb-98bb-9d5d0688e1b1.png)

Browser version | Result
----------------|----------------
Win11 IE 11         | X
Edge 15             | X
Edge 16             | X
Edge 17             | X
Edge 18             | X

**Comment by alexander-akait** [comment]
hm, we can add `window.navigator.msSaveOrOpenBlob` check and if it is `true` we should always run https://github.com/webpack-contrib/worker-loader/blob/master/src/runtime/inline.js#L39

**Comment by elasim** [comment]
- there are security error with using data-uri.
- temporal URL.revokeObjectURL patching isn't useful with core-js is used.
- app side workaround isn't acceptable with some use-case because of heavy workload.

by the way, we probably can merge PR #316 and write caveat about this issue on document or just close that. I'll respect your decision.

**Comment by alexander-akait** [comment]
> there are security error with using data-uri.

What is the problem? Do you mean CORS?

**Comment by elasim** [comment]
> What is the problem? Do you mean CORS?

Umm.. It was `script5022: securityerror`. and it seems like a restriction rather than CORS.
https://docs.microsoft.com/en-us/previous-versions//cc848897(v=vs.85)?redirectedfrom=MSDN

Also, It could be CORS. After some investigation, I've found the information about spec for data-uri and same-origin-policy from stackoverflow. (https://stackoverflow.com/a/23895830)
It said same-origin-policy was changed for data-uri and that wasn't applied to data-uri before.

**Comment by alexander-akait** [comment]
But this problem should be same with blob too

**Comment by elasim** [comment]
> But this problem should be same with blob too

I agree that too. But spec and implementation does.
(IE is annoying and frustating always does 🤣)

Back to solution.

I want to suggest spliting loader with two parts.

first is child compiler which output will be string.
second is code injector which output will be same as current implements.

and if you don't want to make any breaking changes, Perhaps I need to make fork version of this.

**Comment by alexander-akait** [comment]
What is webpack version you are use?

**Comment by elasim** [comment]
I'm using 5.

**Comment by alexander-akait** [comment]
You need this loader only for inline? Because webpack 5 support worker out of box without this loader

**Comment by elasim** [comment]
Yes. This is the only loader can do apply babel-loader for inline-worker as far as I know.
By the way, I'm afraid bothering you because of this. If you are busy or it's annoying you, I'll handle this issue own myself.

**Comment by alexander-akait** [comment]
Honestly because webpack v5 supports workers out of box I think we need new loader, which covert module code to blob, so you can built-in `new Worker('./file.js', import.meta.url)` feature without this loader. 

I think we need deprecate it in near future in favor `new Worker('./file.js', import.meta.url)`, more abilities, more options, better support.
